### ISSUE TITLE
Coverage reporting

### ISSUE THREAD

**Issue by cowboyd** [thefrontside/bigtest#446]
One of the key use-cases is to be able to collect coverage data. In JavaScript, this is done by running the tests against an instrumented build. Because BigTest makes no assumptions about the build, this part needs to be handled by the actuall app.

So if you need to change your webpack config to add instrumentation to your app bundle, then you would have to make that change yourself.

However, once you have an app url that has intrumented code, we can tell bigtest to harvest the coverage data after each test lane is run, and then merge

The convention (in javascript) which is what we're focused on is for instrumentation to write metrics to the `window.__coverage__` object which is a POJO containing all of the coverage data. We can check for this object after the lane run, and send it over the wire to the orchestrator with the `lane:end` event. As each event is received, we can [merge that lane's coverage][1] data into a total coverage object associated with the test run on that agent. The CLI can then take that coverage object and write it to disk in `.bigtest/coverage.json`

In summary

1. each lane will report its coverage
2. lane's coverage will be merged into an aggregate coverage on the `TestRunAgent`
3. CLI can write coverage data to `.bigtest/coverage.json`

[1]: https://github.com/ljharb/istanbul-merge

**Comment by cowboyd** [comment]
For JavaScript code, the coverage library of note is https://github.com/instanbuljs/istanbuljs

If I understand correctly, it appears that what we need to do is to get an istanbul `CoverageMap` object from [`instanbul-lib-coverage`][1]. This is what will be generated inside each lane under the `window.__coverage__` object. Since we cannot depend on agents even having the concept of lanes, it feels most intuitive, for the agent itself to be responsible for assembling a fully merged coverage map for the test run, and only reporting the coverage at the end of its test run.

 That is, we can accumulate an agent level coverage map by merging each coverage map after each lane run. Then, after the entire test run is complete on the agent, we can send the fully merged coverage map with the `testRunAgent:end` event. The orchestrator state can store this in the `coverage` field of of the `TestRunAgentState`.

The alternative would be to send each lane coverage as it is received, but the problem is that it exposes the private structure of the agent's test run into lanes which doesn't feel like a safe assumption to make.

The `coverage` field of the `TestRunAgent` can then be queried via GraphQL in the client so that it can be tranformed on there. So, for example if you specified `--coverage lcov:/var/tmp` it would use the `lcov` formatter to render the coverage data returned by the server to the file `/var/tmp/chrome.lcov` (assuming that chrome was one of the agents). This process would use [`instanbul-lib-report`][2] to convert the raw coverage map into a report of the specified format 

[1]: https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-lib-coverage
[2]: https://github.com/istanbuljs/istanbuljs/tree/master/packages/istanbul-lib-report

<blockquote><img src="https://github.githubassets.com/images/modules/open_graph/github-logo.png" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com">Build software better, together</a></strong></div><div>GitHub is where people build software. More than 50 million people use GitHub to discover, fork, and contribute to over 100 million projects.</div></blockquote>
